nava/functions.py

```diff
@@ -105,16 +105,26 @@ def inner_function(*args, **kwargs):
         :type kwargs: dict
         :return: modified function result
         """
-        sound_path = args[0]
-        if not (isinstance(sound_path, str)):
+        if "sound_path" in kwargs.keys():
+            sound_path = kwargs["sound_path"]
+        elif len(args) >= 1:
+            sound_path = args[0]
+        else: 
```

Comment:
I am thinking about the situation in which this error is going to be raised. The user must call `play()` without any positional nor keyword argument. Isn't that blocked by Python itself to run a function which has a positional argument without any arguments?

If I am missing any other situations please provide examples. :thinking:

Reply:
You are right!
But decorators here run before the primary function.
But we have a better solution, we can define sound_path as a positional argument for decorator something like:
```python
def path_check(func):
    """
    Check the given path to be a string and a valid file directory.

    :return: inner function
    """
    def inner_function(sound_path, *args, **kwargs):
        """
        Inner function.

        :param args: non-keyword arguments
        :type args: list
        :param kwargs: keyword arguments
        :type kwargs: dict
        :return: modified function result
        """
        if not (isinstance(sound_path, str)):
            raise NavaBaseError(SOUND_FILE_PATH_TYPE_ERROR)
        # check sound file existance
        if not (os.path.isfile(sound_path)):
            raise NavaBaseError(SOUND_FILE_EXIST_ERROR)
        return func(sound_path, *args, **kwargs)
    return inner_function
```

@AHReccese please take a look at this possible solution 🔥

Reply:
@sadrasabouri 
Python is a dynamically typed language, which means that the type of a variable can change at runtime. Because of this, Python does not perform strict type-checking during function calls.

If a function is called with input parameters that are not of the expected type, it may result in errors or unexpected behavior at runtime. However, Python does not block such calls because it allows developers to write more flexible and expressive code.

Reply:
@sepandhaghighi 
I know how you feel here, but don't you think adding sound_path next to args and kwargs kinda vanishes the meaning of the `argument` concept?

Reply:
I create a simple function(I'm not a much fan of it, but still) to avoid duplication(although duplication is meant to be more or equal to 3 similar occurrences according to a known rule of thumb).
@sepandhaghighi @sadrasabouri

Reply:
> @sepandhaghighi I know how you feel here, but don't you think adding sound_path next to args and kwargs kinda vanishes the meaning of the `argument` concept?

BTW, I think @sepandhaghighi 's solution is better here. With yours we are now somehow emulating the `kwargs` and `args` behavior and I don't like that. Plus that it's not very expandable comparing to Sepand's. While you'd need to add a positional argument to the `inner_function` for new argument in Sepand's solution, you should define an extractor function in yours.

Reply:
@sadrasabouri 
how about this? 
```{python}
sound_path = kwargs.get('sound_path', None if len(args) == 0 else args[0]) 
if sound_path is None:
    raise NavaBaseError(SOUND_FILE_PATH_PARAM_ERROR)
return sound_path
```

Reply:
@sepandhaghighi @sadrasabouri 
I applied Sepand's suggestion.

Reply:
> @sepandhaghighi @sadrasabouri I applied Sepand's suggestion.

Perfect!